src/lucas_kanade.jl

```diff
@@ -1,259 +1,255 @@
-        # Calculate Ixx, Iyy and Ixy taking Gaussian weights for pixels in the grid.
-        # Return integral tables to facilitate fast sums over regions in Ixx, Iyy and Ixy.
-        Ixx_integral_table, Ixy_integral_table, Iyy_integral_table =  compute_partial_derivatives(Ix, Iy)
-
-        inner_bounds = map(i -> first(i) +  algorithm.window_size:last(i) -  algorithm.window_size, axes(first_pyramid[i]))
-        # Running the algorithm for each input point.
-        for j = 1:length(displacement)
-            if status[j]
-                # Position of the point in the current pyramid level.
-                point = determine_pyramid_coordinates(points, i, j)
-                # If the window falls inside the first image we pre-compute the
-                # spatial gradient matrix because in many instance the window
-                # will also be completely contained inside the second image.  If
-                # it turns out that the  corresponding window is not totally
-                # contained inside the second window then we will recompute the
-                # spatial gradient matrix again later.
-                if lies_in(inner_bounds, point)
-                    square_grid = SVector{2}(point[1] -  algorithm.window_size:point[1] +  algorithm.window_size, point[2] -  algorithm.window_size:point[2] +  algorithm.window_size)
-                    G = compute_spatial_gradient(square_grid, Ixx_integral_table, Ixy_integral_table, Iyy_integral_table)
+function optflow!(
+    first_img::AbstractMatrix{T}, second_img::AbstractMatrix{T},
+    points::Vector{SVector{2, Float64}}, displacement::Vector{SVector{2, Float64}},
+    algorithm::LucasKanade,
+) where T <: Gray
+    first_pyramid = LKPyramid(first_img, algorithm.pyramid_levels)
+    second_pyramid = LKPyramid(
+        second_img, algorithm.pyramid_levels; compute_gradients=false,
+    )
+    optflow!(first_pyramid, second_pyramid, points, displacement, algorithm)
+end
+
+function optflow!(
+    first_pyramid::LKPyramid, second_pyramid::LKPyramid,
+    points::Vector{SVector{2, Float64}}, displacement::Vector{SVector{2, Float64}},
+    algorithm::LucasKanade,
+)
+    has_enough_layers =
+        length(first_pyramid.layers) > algorithm.pyramid_levels &&
+        length(second_pyramid.layers) > algorithm.pyramid_levels
+    has_enough_layers || throw("Not enough layers in pyramids.")
+
+    n_points = length(points)
+    status = trues(n_points)
+
+    window = algorithm.window_size
+    window2x = 2 * window
+
+    for level in (algorithm.pyramid_levels + 1):-1:1
+        level_resolution = axes(first_pyramid.layers[level])
+        # Interpolate layer to get sub-pixel precision.
+        # We never go out-of-bound, so there is no need to extrapolate.
+        interploated_layer = interpolate(second_pyramid.layers[level], BSpline(Linear()))
+
+        for did in 1:n_points
```

Comment:
Is there a particular reason why you called the loop variable `did`? I keep reading it as the past tense of "to do" and it feels a little odd. Perhaps a vanilla `n` would work better here?

Reply:
It originally standed for `displacement_id`, but I agree, it is confusing. Changed to `n`.